plugin/acl/setup.go

```diff
@@ -0,0 +1,188 @@
+package acl
+
+import (
+	"net"
+	"strings"
+
+	"github.com/coredns/coredns/core/dnsserver"
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/plugin/metrics"
+
+	"github.com/caddyserver/caddy"
+	"github.com/infobloxopen/go-trees/iptree"
+	"github.com/miekg/dns"
+)
+
+const (
+	// QtypeAll is used to match any kinds of DNS records type.
+	// NOTE: The value of QtypeAll should be different with other QTYPEs defined in miekg/dns.
+	QtypeAll uint16 = 299
+)
+
+var (
+	// DefaultFilter is the default ip filter which matches both IPv4All and IPv6All.
+	DefaultFilter *iptree.Tree
+)
+
+func init() {
+	initDefaultFilter()
+
+	caddy.RegisterPlugin("acl", caddy.Plugin{
+		ServerType: "dns",
+		Action:     setup,
+	})
+}
+
+func initDefaultFilter() {
+	_, IPv4All, _ := net.ParseCIDR("0.0.0.0/0")
+	_, IPv6All, _ := net.ParseCIDR("::/0")
+	DefaultFilter = iptree.NewTree()
+	DefaultFilter.InplaceInsertNet(IPv4All, struct{}{})
+	DefaultFilter.InplaceInsertNet(IPv6All, struct{}{})
+}
+
+func setup(c *caddy.Controller) error {
+	a, err := parse(c)
+	if err != nil {
+		return plugin.Error("acl", err)
+	}
+
+	dnsserver.GetConfig(c).AddPlugin(func(next plugin.Handler) plugin.Handler {
+		a.Next = next
+		return a
+	})
+
+	// Register all metrics.
+	c.OnStartup(func() error {
+		metrics.MustRegister(c, RequestBlockCount, RequestAllowCount)
+		return nil
+	})
+	return nil
+}
+
+func parse(c *caddy.Controller) (acl, error) {
+	a := acl{}
+	for c.Next() {
+		r := Rule{}
+		r.Zones = c.RemainingArgs()
+		if len(r.Zones) == 0 {
+			// if empty, the zones from the configuration block are used.
+			r.Zones = make([]string, len(c.ServerBlockKeys))
+			copy(r.Zones, c.ServerBlockKeys)
+		}
+		for i := range r.Zones {
+			r.Zones[i] = plugin.Host(r.Zones[i]).Normalize()
+		}
+
+		for c.NextBlock() {
+			p := Policy{}
+
+			action := strings.ToLower(c.Val())
+			if action == "allow" {
+				p.action = ALLOW
```

Comment:
s/ALLOW/Allow/g

Using all caps to signify constants isn't a Go idiom and is not done in the rest of CoreDNS